File: backend/src/main/java/com/sdsu/backend/service/UserService.java
package com.sdsu.backend.service;

import com.sdsu.backend.model.User;
// import com.sdsu.backend.model.UserSettings;
import com.sdsu.backend.repository.UserRepository;
// import com.sdsu.backend.repository.UserSettingsRepository;
import jakarta.transaction.Transactional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.*;

@Service
public class UserService {

    private final UserRepository userRepository;
    // private final UserSettingsRepository userSettingsRepository;

    // UserSettingsRepository userSettingsRepository (REMOVED THIS PARAMETER FROM
    // USERSERVICE PUT BACK AFTER)
    @Autowired
    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
        // this.userSettingsRepository = userSettingsRepository;
    }

    /*
    public User createUser(String email, String password, String name) {
        User user = new User(email, password, name);
        // UserSettings defaultSettings = new UserSettings(user, false, false);
        // user.setUserSettings(defaultSettings);

        return userRepository.save(user);
    }
     */

    public User save(User user) {
        if (user == null) {
            throw new IllegalArgumentException(("User cannot be null"));
        }
        if (user.getEmail() == null) {
            throw new IllegalArgumentException(("User must have an Email"));
        }
        if (user.getUsername() == null || user.getPassword() == null) {
            throw new IllegalArgumentException(("User must have both a Username and/or Password"));
        }

        return userRepository.save(user);
    }

    /*
    // Fetch All Users
    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    // Fetch User by ID
    public Optional<User> getUserById(Long id) {
        return userRepository.findById(id);
    }

     */

    // Find User by ID
    public Optional<User> findById (Long id) { return userRepository.findById(id); }

    // Find User by Email
    public Optional<User> getUserByEmail(String email) {
        return userRepository.findUserByEmail(email);
    }

    // Find User by Password
    public Optional<User> getUserByPassword(String password) {
        return userRepository.findUserByPassword(password);
    }

    // Find User by Name
    public Optional<User> getUserByName(String name) {
        return userRepository.findUserByName(name);
    }

    // Find All Active Users
    public List<User> getUserByActiveStatus(boolean activeStatus) {
        return userRepository.findUserByActiveStatus(activeStatus);
    }

    // Update User Information
    public User updateUser(Long id, String email, String password, String name) {
        User user = userRepository.findById(id).orElseThrow(() -> new RuntimeException("User not found"));
        if (email != null)
            user.setEmail(email);
        if (password != null)
            user.setPassword(password);
        if (name != null)
            user.setUsername(name);

        return userRepository.save(user);
    }

    // Delete User
    public void deleteUser(Long id) {
        userRepository.deleteById(id);
    }

    @Transactional
    public void deleteUserByEmail(String email) {
        if (email == null) {
            throw new IllegalArgumentException("Email cannot be null to be delete");
        }
        userRepository.deleteByEmail(email);
    }
}

File: frontend/src/services/userService.js
// Base URL for your Spring Boot backend
const API_BASE_URL = 'http://localhost:8080/api';

// ===== CREATE A NEW USER =====

export const createUser = async (userData) => {
    try {
        const response = await fetch(`${API_BASE_URL}/users`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                email: userData.email,
                password: userData.password,
                name: userData.name,
            })
        });

        if (!response.ok) {
            throw new Error('Failed to create user');
        }

        return await response.json();
    } catch (error) {
        console.error('Error creating assignment:', error);
        throw error;
    }
};

// ===== READ - Get User's email/password/username =====

// Get All the Users
export const getAllActiveUsers = async () => {
    try {
        const response = await fetch(
            `${API_BASE_URL}/users?activeStatus=true`
        );

        if (!response.ok) {
            throw new Error('Failed to fetch all Active Users');
        }

        return await response.json();
    } catch (error) {
        console.error('Error fetching Active Users:', error);
        throw error;
    }
}

// Get User via Email

export const getUserByEmail = async (email) => {
    try {
        const response = await fetch(
            `${API_BASE_URL}/users/email/${email}`
        );

        if (!response.ok) {
            throw new Error('Failed to fetch a User with this EMAIL');
        }

        return await response.json();
    } catch (error) {
        console.error('Error fetching User via this EMAIL', error);
        throw error;
    }
}

// Get User via Password
export const getUserByPassword = async (password) => {
    try {
        const response = await fetch(
            `${API_BASE_URL}/users/password/${password}`
        );

        if (!response.ok) {
            throw new Error('Failed to fetch a User with this PASSWORD');
        }

        return await response.json();
    } catch (error) {
        console.error('Error fetching User via this PASSWORD', error)
    }
}

// Get User via Name
export const getUserByName = async (name) => {
    try {
        const response = await fetch(
            `${API_BASE_URL}/users/name/${name}`
        );

        if (!response.ok) {
            throw new Error('Failed to fetch a User with this NAME');
        }

        return await response.json();
    } catch (error) {
        console.error('Error fetching User via this NAME', error)
    }
}


// ===== UPDATE - Edit User's information =====
// Update via Email
export const updateUser = async (email, userData) => {
    try {
        const response = await fetch(`${API_BASE_URL}/users/email/${email}`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                email: userData.email,
                password: userData.password,
                name: userData.name,
            }),
        });

        if (!response.ok) {
            throw new Error('Failed to update this user');
        }

        return await response.json();
    } catch (error) {
        console.error('Error updating USER', error);
        throw error;
    }
}

// ===== DELETE - Remove A User =====
// Delete a User via Email
/*
export const deleteUser = async (email) => {
    try {
        const response = await fetch(`${API_BASE_URL}/users/email/${email}`, {
            method: 'DELETE'
        })
    }
}

 */



